File: typescript/tool.ts
import * as fs from 'fs';

defineAst("", "Expr", [
    "Binary   : Expr left, Token operator, Expr right",
    "Grouping : Expr expression",
    "Literal  : Object value",
    "Unary    : Token operator, Expr right"]);

function defineAst(outputDir: string, baseName: string, types: string[]) {
    let data = '';
    data += "import {Token} from \"./token\";";
    data += "\n\n";
    // data += `export abstract class ${baseName} {}`;
    let typeNames: string[] = [];
    types.forEach(t => {
        typeNames.push(t.split(":")[0].trim());
    });
    data += `export type ${baseName} = ${typeNames.join(" | ")}`;
    data += `\n\n`;

    for (const t of types) {
        const className = t.split(":")[0].trim();
        const fields = t.split(":")[1].trim();
        data = defineType(data, baseName, className, fields);
    }

    data = defineVisitor(data, baseName, types);

    fs.writeFileSync(`./interpreter/${baseName.toLowerCase()}.ts`, data, {encoding: 'utf8', flag: 'w'});
    console.log('File written successfully.');
}

function defineType(data: string, baseName: string, className: string, fieldsStr: string): string {
    // data += `export class ${className} extends ${baseName} {\n`;
    data += `export class ${className} {\n`;
    data += `    constructor(`;
    const fields = fieldsStr.split(',');
    for (const f of fields) {
        const id = f.trim().split(" ")[1];
        const t = f.trim().split(" ")[0];
        data += `public ${id}: ${t}, `;
    }

    data = data.substring(0, data.length - 2);
    // data += `){ super() }\n`;
    data += `){}\n`;
    data += `\n`;
    data += `    public accept<T>(visitor: Visitor<T>): T{\n`;
    data += `        return visitor.visit${className}${baseName}(this);\n`;
    data += `    }\n`;
    data += `}`;
    data += "\n\n";
    return data;
}

function defineVisitor(data: string, baseName: string, types: string[]): string {
    data += `export interface Visitor<T> {\n`;
    types.forEach((type: string) => {
        const className = type.split(":")[0].trim();
        data += `    visit${className}${baseName}(${baseName.toLowerCase()}: ${className}): T\n`;
    });
    data += `}\n`;
    return data;
}

File: pylox/tools/tool.py
import os


def define_ast(output_dir: str, base_name: str, types: list[str]):
    absolute = os.path.dirname(__file__)
    path = os.path.join(absolute + output_dir, base_name.lower() + ".py")
    if os.path.exists(path):
        os.remove(path)
    file = ""
    for t in types:
        class_name = t.split(":")[0].strip()
        fields = t.split(":")[1].strip()
        file = define_class(file, class_name, base_name, fields)
    with open(os.path.join(absolute + output_dir, base_name.lower() + ".py"), "w") as f:
        f.write(file)


def define_class(file, class_name, base_name, fields):
    file += "\n\n"
    fields = fields.split(",")
    types = list()
    identifiers = list()
    for field in fields:
        types.append(field.strip().split(" ")[0])
        identifiers.append(field.strip().split(" ")[1])

    file += "class " + class_name + ":\n"
    ctor = "    def __init__(self"
    assign = ""
    zipped = zip(types, identifiers)
    for itr in zipped:
        ctor += f", {itr[1]}"
        assign += f"        self.{itr[1]} = {itr[1]}\n"
    ctor += "):\n"
    file += ctor
    file += assign
    file += "\n"
    file += "    def accept(self, visitor):\n"
    file += (
        f"        return visitor.visit_{class_name.lower()}_{base_name.lower()}(self)\n"
    )
    file += f"\n    def __eq__(self, other):\n"
    file += f"        return (\n"
    for id in identifiers:
        file += f"            self.{id} == other.{id} and\n"
    file = file[:-4]
    file += "\n        )"
    return file


if __name__ == "__main__":
    output_dir = ""
    define_ast(
        output_dir,
        "Expr",
        [
            "Assign   : Token name, Expr value",
            "Binary   : Expr left, Token operator, Expr right",
            "Grouping : Expr expression",
            "Literal  : Object value",
            "Unary    : Token operator, Expr right",
            "Variable : Token name",
        ],
    )
    define_ast(
        output_dir,
        "Stmt",
        [
            "Block      : list[Stmt] statements",
            "Expression : Expr expression",
            "Print      : Expr expression",
            "Var        : Token name, Expr initializer",
        ],
    )
